Cycle with super traits and negative bounds

I tried this code:

```rust
#![feature(negative_bounds)]

trait A: !B {}
trait B: !A {}

fn main() {}
```

I expected to see this happen: Should compile (because this isn't a cycle due to the negative bound).

Instead, this happened: errors with

```
error[E0391]: cycle detected when computing the super predicates of `A`
 --> m.rs:3:10
  |
3 | trait A: !B {}
  |          ^^
  |
note: ...which requires computing the super predicates of `B`...
 --> m.rs:4:10
  |
4 | trait B: !A {}
  |          ^^
  = note: ...which again requires computing the super predicates of `A`, completing the cycle
note: cycle used when collecting item types in top-level module
 --> m.rs:3:1
  |
3 | trait A: !B {}
  | ^^^^^^^^^^^^^^
```

### Meta

`rustc --version --verbose`:
```
rustc 1.71.0-nightly (473f916d8 2023-05-03)
binary: rustc
commit-hash: 473f916d836cc662c5bdbb0d40af9fb4678fab9e
commit-date: 2023-05-03
host: aarch64-apple-darwin
release: 1.71.0-nightly
LLVM version: 16.0.2
```

I am fully aware that this feature just landed for internal experimentation only. So if this bug isn't an actual bug or is simply out of scope, feel free to close this issue :)

--- a/compiler/rustc_hir_analysis/src/collect/predicates_of.rs
+++ b/compiler/rustc_hir_analysis/src/collect/predicates_of.rs
@@ -657,14 +657,15 @@ pub(super) fn implied_predicates_with_filter(
         &*tcx.arena.alloc_from_iter(superbounds.predicates().chain(where_bounds_that_match));
     debug!(?implied_bounds);
 
-    // Now require that immediate supertraits are converted,
-    // which will, in turn, reach indirect supertraits.
+    // Now require that immediate supertraits are converted, which will, in
+    // turn, reach indirect supertraits, so we detect cycles now instead of
+    // overflowing during elaboration.
     if matches!(filter, PredicateFilter::SelfOnly) {
-        // Now require that immediate supertraits are converted,
-        // which will, in turn, reach indirect supertraits.
         for &(pred, span) in implied_bounds {
             debug!("superbound: {:?}", pred);
-            if let ty::PredicateKind::Clause(ty::Clause::Trait(bound)) = pred.kind().skip_binder() {
+            if let ty::PredicateKind::Clause(ty::Clause::Trait(bound)) = pred.kind().skip_binder()
+                && bound.polarity == ty::ImplPolarity::Positive
+            {
                 tcx.at(span).super_predicates_of(bound.def_id());
             }
         }

--- a/tests/ui/traits/negative-bounds/supertrait.rs
+++ b/tests/ui/traits/negative-bounds/supertrait.rs
@@ -1,0 +1,9 @@
+// check-pass
+
+#![feature(negative_bounds)]
+//~^ WARN the feature `negative_bounds` is incomplete
+
+trait A: !B {}
+trait B: !A {}
+
+fn main() {}

--- a/tests/ui/traits/negative-bounds/supertrait.stderr
+++ b/tests/ui/traits/negative-bounds/supertrait.stderr
@@ -1,0 +1,10 @@
+warning: the feature `negative_bounds` is incomplete and may not be safe to use and/or cause compiler crashes
+  --> $DIR/supertrait.rs:3:12
+   |
+LL | #![feature(negative_bounds)]
+   |            ^^^^^^^^^^^^^^^
+   |
+   = note: `#[warn(incomplete_features)]` on by default
+
+warning: 1 warning emitted
+

